src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1430,10 +1430,9 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
         continue;
       }
       failureMessage = canPassThroughCanal(canalAttachment, units, player, data);
-      if (!Properties.getControlAllCanalsBetweenTerritoriesToPass(data) && !failureMessage.isPresent()) {
-        return Optional.empty();
-      } else if (Properties.getControlAllCanalsBetweenTerritoriesToPass(data) && failureMessage.isPresent()) {
-        return failureMessage;
+      if ((!Properties.getControlAllCanalsBetweenTerritoriesToPass(data) && !failureMessage.isPresent())
```

Comment:
This took a step in the wrong direction IMO. Root issue I see is SRP violation with `failureMessage`, has one value with one meaning in one case, and another value in another. Specifically one thing that is confusing is that when it is empty, it contradicts the name, since an Optional.empty is not a "failureMessage". There is also some magic with state of "failureMessage" and the decision we make when we hit the conditional.

How about something more like this?

```
  if (!isCanalOnRoute(canalAttachment, route, data)) {
         // filter to just canals that are on the route.  
          continue;
        }
if( canPassThroughCanal(canalAttachment, units, player, data) && !Properties.getControlAllCanalsBetweenTerritoriesToPass(data) ) {
    return Optional.empty()
}
if( Properties.getControlAllCanalsBetweenTerritoriesToPass(data) && ! canPassThroughCanal(canalAttachment, units, player, data) ) {
   Optional<String> failMsg = canPassThroughCanal(canalAttachment, units, player, data);
   return failMsg;
}
```